Plans: Add plan to cart and redirect to "add domain" for special query string

sirbrillig opened the pull request:
#### Changes proposed in this Pull Request

There is a sidebar upsell that is displayed for sites with no plan that reads "Free domain with an annual plan" with button that reads "Upgrade". The intent of this button is to start a flow where the customer purchases a plan and a domain.

The way in which this currently works is that the upsell redirects you to the "add domain" page with a flag that silently adds a Personal plan to your cart as soon as the page loads. This can cause a lot of confusion if the customer then does not select a domain and navigates away because the plan has already been added to the cart which will trigger cart abandonment notices and emails even though the customer didn't intend to add that plan.

The PR adds a flag to the plans page which will cause the plan purchase buttons to add the selected plan to your cart and then redirect to the "add domain" page. This way adding the plan becomes explicit and also provides the customer the opportunity to choose which plan they would prefer.

This does not remove the flag on the domains page yet, although it can be removed as soon as D59605-code is merged.

Fixes https://github.com/Automattic/wp-calypso/issues/51212

Includes https://github.com/Automattic/wp-calypso/pull/51605 and will require a rebase when that is merged.

This requires D59605-code to modify where the upsell button goes, but can be safely merged before that.

<img width="275" alt="Screen Shot 2021-04-01 at 6 02 43 PM" src="https://user-images.githubusercontent.com/2036909/113358817-7ebee200-9314-11eb-925f-c470ba52598f.png">

<img width="1064" alt="Screen Shot 2021-04-03 at 5 30 55 PM" src="https://user-images.githubusercontent.com/2036909/113491942-64b30a00-94a2-11eb-82bd-1434133b4e7c.png">

<img width="910" alt="Screen Shot 2021-04-01 at 6 02 27 PM" src="https://user-images.githubusercontent.com/2036909/113358849-88484a00-9314-11eb-8fc5-722234ddd535.png">


To do:

- [x] Hide the monthly plans if this flag is set because they do not come with a domain.
- [x] Verify that this works (or at least doesn't cause problems) for Jetpack sites.
- [x] Maybe hide that "If you upgrade to the Personal plan" upsell on the "add domain" page if the cart contains a plan. (Turns out that was a bug; it was already supposed to do that.)
- [x] Maybe change the "Upgrade" links on the plans page to read "Add to cart" as described in https://github.com/Automattic/wp-calypso/issues/51212#issuecomment-810939074 (this also exposed a bug which I've fixed here.)
- [x] Maybe remove the old `ref=upgrade` behavior in a later PR to make the transition easier between the backend change and the frontend. (Moved to https://github.com/Automattic/wp-calypso/pull/51642)

#### Testing instructions

- Apply D59605-code and sandbox the API.
- Start with a site that has no paid plan.
- Visit the customer home page for the site, eg: http://calypso.localhost:3000/home/example.com
- Click the "Upgrade" button in the sidebar notification that reads "Free domain with an annual plan" (see first screenshot above).
- Verify that you are redirected to the plans page for your site (see second screenshot above).
- Click the "Upgrade" button on a plan's card to select that plan.
- Verify that you are redirected to the "Site Domains" page (see third screenshot above).
- Click to select a domain.
- Click through any email upsell you may see.
- Verify that you arrive at checkout with both the plan and the domain in your cart.

sirbrillig committed:
Refactor plan upgrade redirect to create checkout url just in time

This will allow us to make all decisions related to checkout redirect in
one place rather than making some of those decisions in the Redux
connect function. As a result we're able to change those decisions more
easily.

sirbrillig committed:
Wrap Plans page in CalypsoShoppingCartProvider

sirbrillig committed:
Add plan to cart and visit domains for redirectToAddDomainFlow prop

sirbrillig committed:
Activate redirectToAddDomainFlow with domainAddFlow query on plans

michaeldcain commented:
> Verify that this works (or at least doesn't cause problems) for Jetpack sites.

I tested with a jurassic.ninja site. The nudge doesn't show for Jetpack sites, but manually navigating to the Jetpack plans page with the query string doesn't change the current behavior of the plans page. It _does_ persist the query string through to Checkout, so we might want to strip that, but I don't know if it's really worth it.

<img width="537" alt="Screen Shot 2021-04-01 at 11 27 50 AM" src="https://user-images.githubusercontent.com/942359/113317641-9169f480-92dd-11eb-819b-b41c6faea70f.png">

sirbrillig committed:
Hide monthly options when in redirectToAddDomainFlow

sirbrillig committed:
Fix: hide NewDomainsRedirectionNoticeUpsell when plan in cart

sirbrillig committed:
Fix ternary logic in PlanFeaturesActionsButton

I think the intent was to say "use the buttonText from props, or
determine the button text based on if the plan is free", but the first
two conditions were combined so instead the behavior was "if there is
buttonText in the props or if the plan is free, display the free plan
text and otherwise display the upgrade text".

This separates the two conditions so it should be more clear what
happens when.

sirbrillig committed:
Override the PlanFeatures button text if in domain redirect flow

This makes it so that the plan purchase buttons all read "Add to Cart"
if we are in the plan + domain flow.

sirbrillig committed:
Change query string to addDomainFlow

michaeldcain reviewed (approved):
This works as expected with the backend patch applied, and is overall a huge improvement. Although I think we could go ahead and ship it (given the failing tests are unrelated), I do have the following feedback:

I know I approved the backend diff that links to `/plans/{siteSlug}?addDomainFlow=true`, but after using the flow, I wonder if it would still be better (and more future proof) to link directly to the `/plans/yearly/` page. As it currently is, the layout of the plans page (and even the tab filter's purpose) changes for a user depending on if they click on the nudge or not:

![plan-upgrade](https://user-images.githubusercontent.com/942359/113481510-f5202900-9467-11eb-8b0e-bcc96710c74b.gif)

This can be confusing, and implies that we only bundle the domain with the "blogs and personal sites" plans (if the user even notices the tab change). I'd personal prefer a 4-up plan list with the tabs hidden entirely.

It's also worth noting that as Marketing rolls out monthly/annual upgrades/downgrades, there's the possibility that `eligibleForWpcomMonthlyPlans` might go away, in which case this flow would break. At minimum, we might want to add a todo to the code to remind people to update the nudge in that case.

What are your thoughts?

For anyone without a sandbox, here's a gif of the updated flow:

![upgrade-flow](https://user-images.githubusercontent.com/942359/113481835-d884f080-9469-11eb-9305-8bf204ebf17f.gif)

sirbrillig commented:
I agree that the experience feels better with the full list of plans, but the whole system in `PlansFeaturesMain` through `PlanFeatures` and into `PlanFeaturesHeader` seems to be based on [two options](https://github.com/Automattic/wp-calypso/blob/f0896401b64320191b0ce015957192d12e2966d2/client/my-sites/plans-features-main/index.jsx#L505): `interval` (the full list of plans with a monthly/yearly tab bar) or `customer` (the yearly plans split with a person/business tab bar). I'm not clear what the `/plans/yearly` URL does, since for me it still shows the monthly plan selector:

<img width="955" alt="Screen Shot 2021-04-03 at 5 17 52 PM" src="https://user-images.githubusercontent.com/2036909/113491696-92974f00-94a0-11eb-823e-57b81e7f8162.png">

I've experimented a bit with adding a third option: `yearly`, but I haven't yet figured out how to get it to change the tabs appropriately. I'll keep trying, but if you know of some way to get the view you're describing without having to modify all of the plans view components I'm all ears!

sirbrillig committed:
Hide plan selector but use interval type for redirectToAddDomainFlow

sirbrillig commented:
I did figure out a way to hide the plan selector! Hopefully that will work (and hopefully it won't show the monthly list first for some reason... the way the plans list works is very complex).

<img width="1064" alt="Screen Shot 2021-04-03 at 5 30 55 PM" src="https://user-images.githubusercontent.com/2036909/113491960-90ce8b00-94a2-11eb-8833-af7db99dd528.png">

michaeldcain commented:
> I'm not clear what the /plans/yearly URL does, since for me it still shows the monthly plan selector

Sorry to send you down that rabbit hole. I don't see it in the controller for the section, so it's probably just hacked in there. I assumed that there were explicit routes to default to either the monthly or yearly plans, since the tab selector seems to link to them:

![monthly-yearly](https://user-images.githubusercontent.com/942359/113579355-2fb5cd00-95f2-11eb-91b1-e26e1afc1cdb.gif)

Regardless, let's just ignore that suggestion :) - this whole section seems a little overly complex. Your latest change to show the 4-up plans page and hide the tab works really well. I tested the flow on a qualifying site (works perfectly), a Jetpack site (no change), and an unqualified site with a plan by manually adding the query string. The latter has some wonkiness where only the available plans have CTAs (and the current plan loses its "manage" button), but should theoretically not happen if the JITM system is working:

<img width="1447" alt="Screen Shot 2021-04-05 at 9 44 35 AM" src="https://user-images.githubusercontent.com/942359/113580468-c9ca4500-95f3-11eb-8dec-7e396f472d20.png">

Unless that last one is a quick fix, I think we should just go ahead and ship this - it's a much better improvement over the current flow.

sirbrillig committed:
Always display "Add to Cart" button if redirectToAddDomainFlow active

sirbrillig commented:
> Unless that last one is a quick fix, I think we should just go ahead and ship this - it's a much better improvement over the current flow.

Good eye. 🤔  This is why I was afraid of messing with the can of worms that is the plans page. This appears to be happening because we're forcing the `interval` view type on a page that otherwise wouldn't show that type; I don't fully understand why but the code for that view explicitly hides the button entirely for a plan that cannot be purchased and isn't the current plan. I've pushed a commit to force the button to be visible when in the `addDomainFlow`, which solves that issue but it does make me worry about what other issues it might cause. In any case, this should be good enough for now.

sirbrillig committed:
Wrap PlanFeatures directly in CalypsoShoppingCartProvider

This component is used in so many places it's hard to be sure otherwise
that all the places have a provider available.

sirbrillig committed:
Remove CalypsoShoppingCartProvider from Plans page

sirbrillig commented:
Thanks to the e2e tests (🎉 ) I noticed that this broke a number of other locations that used the `PlanFeatures` component and were not inside a `ShoppingCartProvider`. I've updated the PR to wrap the component itself in a provider. It's not ideal, but `PlanFeatures` is used in quite a lot of places and this seems the only prudent way at the moment to make sure we always have a cart manager available.
